source/features/useful-not-found-page.tsx

```diff
@@ -29,71 +27,247 @@ async function checkAnchor(anchor: HTMLAnchorElement): Promise<void> {
+	const commitDate = lastCommitInfo.commit.committer.date;
+	const linkToCommit = lastCommitInfo.html_url;
+
+	url.assign({
+		route: 'commits'
+	})
+	const linkToCommitHistory = url.toString();
+
+	if (fileInfo.status === 'removed') {
+		const linkToLastVersion = fileInfo.blob_url;
+
+		return {
+			type: 'removed',
+			filePath,
+			commitSha,
+			commitDate,
+			linkToCommit,
+			linkToLastVersion,
+			linkToCommitHistory,
+		};
+	}
+
+	if (fileInfo.status === 'renamed') {
+		const newFilePath = fileInfo.filename;
+		filePath = fileInfo.previous_filename;
+
+		const linkToNewVersion = fileInfo.blob_url;
+		url.assign({
+			route: 'blob',
+			branch: lastCommitInfo.parents[0].sha,
+			filePath
+		})
+		let linkToLastVersion = url.toString();
+
+		return {
+			type: 'renamed',
+			filePath,
+			newFilePath,
+			commitSha,
+			commitDate,
+			linkToCommit,
+			linkToLastVersion,
+			linkToNewVersion,
+			linkToCommitHistory,
+		};
+	}
 }
 
-// If the resource exists in the default branch, link to it
-async function addDefaultBranchLink(bar: Element): Promise<void> {
-	const parts = getCleanPathname().split('/');
-	const branch = parts[3];
-	if (!branch) {
+async function getLinkToCommitHistoryOnDefaultBranch(): Promise<string | undefined> {
+	// Get the current branch
+	const url = new GitHubURL(location.href);
+	const currentBranch = url.branch;
+
+	if (!currentBranch) {
 		return;
 	}
 
 	const defaultBranch = await getDefaultBranch();
-	if (branch === defaultBranch) {
+	if (currentBranch === defaultBranch) {
 		return;
 	}
 
-	parts[3] = defaultBranch; // Change branch
-	const url = '/' + parts.join('/');
-	if (await is404(location.origin + url)) {
+	url.assign({
+		route: 'commits',
+		branch: defaultBranch,
+	})
+	// Check if that path exists
+	if (await is404(url.toString())) {
 		return;
 	}
 
-	bar.after(
-		<p className="container mt-4 text-center">
-			The {getType()} exists on the <a href={url}>default branch</a>.
-		</p>,
-	);
+	// If it does, return it
+	return url.toString();
 }
 
-function init(): false | void {
-	const parts = parseCurrentURL();
-	if (parts.length <= 1 || !select.exists('[alt*="This is not the web page you are looking for"]')) {
-		return false;
+async function whatHappenedToTheFile(): Promise<Record<string, string | undefined> | void> {
+	// First, check if the file exists on the default branch
+	const urlOnDefaultBranch = await getUrlToFileOnDefaultBranch();
+	if (urlOnDefaultBranch) {
+		return {
+			type: 'isOnDefaultBranch',
+			linkToLastVersion: urlOnDefaultBranch,
+		};
 	}
 
-	const bar = <h2 className="container mt-4 text-center"/>;
+	// Else, check if the file exists in commit history
+	const lastCommitIncludingFile = await getLastCommitForFile();
+	if (lastCommitIncludingFile) {
+		return lastCommitIncludingFile;
+	}
+
+	// Else, check if the file exists in the default branch's history
+	const linkToCommitHistoryOnDefaultBranch
+		= await getLinkToCommitHistoryOnDefaultBranch();
```

Comment:
This seems identical to `getLastCommitForFile`, in practice, except that it's for a different branch. We just need to verify whether it ever existed (via the single v3 call I mentioned in the other comment). I think the whole `getLinkToCommitHistoryOnDefaultBranch` can be dropped.